java/src/Ice/src/main/java/com/zeroc/Ice/UnknownException.java

```diff
@@ -1,42 +1,24 @@
-//
-// Copyright (c) ZeroC, Inc. All rights reserved.
-//
-//
-// Ice version 3.7.10
-//
-// <auto-generated>
-//
-// Generated from file `LocalException.ice'
-//
-// Warning: do not edit this file.
-//
-// </auto-generated>
-//
+// Copyright (c) ZeroC, Inc.
 
 package com.zeroc.Ice;
 
 /**
- * This exception is raised if an operation call on a server raises an unknown exception. For
- * example, for C++, this exception is raised if the server throws a C++ exception that is not
- * directly or indirectly derived from <code>Ice::LocalException</code> or <code>Ice::UserException
- * </code>.
+ * This exception is raised if an operation call on a server raises an unknown exception, that is,
+ * any exception which does not derive from {@link LocalException} or {@link UserException}.
  */
 public class UnknownException extends LocalException {
-  public UnknownException() {
-    this.unknown = "";
-  }
-
   public UnknownException(Throwable cause) {
```

Comment:
Java makes frequent use of the constructor, mostly in cases where we encounter a competely unknown exception, that we know nothing else about, and just wrap in an `UnknownException`.
Looking at the corresponding code in C#, we don't handle these exceptions, and just let them throw upwards.
But, some of those places have comments which suggests the Java behavior is more correct.

For example in `ConnectionI`, java has an extra `catch` for truly unknown exceptions:
```
} catch (LocalException ex) {
  dispatchException(ex, requestCount);
} catch (RuntimeException | java.lang.Error ex) {
  // A runtime exception or an error was thrown outside of servant code (i.e., by Ice code).
  // Note that this code does NOT send a response to the client.
  var uex = new UnknownException(ex);
```
and in C#, we just have a TODO.
```
catch (LocalException ex) // TODO: catch all exceptions
{
    dispatchException(ex, requestCount);
}
```

----

Or in `LoggerAdminLoggerI`, java has an extra `else` to handle truly unknown exceptions,
```
if (ex instanceof com.zeroc.Ice.CommunicatorDestroyedException) {
  // Expected if there are outstanding calls during communicator destruction.
} else if (ex instanceof com.zeroc.Ice.LocalException) {
  _loggerAdmin.deadRemoteLogger(
      p, _localLogger, (com.zeroc.Ice.LocalException) ex, "log");
} else {
  _loggerAdmin.deadRemoteLogger(
      p, _localLogger, new com.zeroc.Ice.UnknownException(ex), "log");
}
```
whereas C# looks like it just silently ignores unknown errors here:
```
catch (AggregateException ae)
{
    if (ae.InnerException is Ice.CommunicatorDestroyedException)
    {
        // expected if there are outstanding calls during communicator destruction
    }
    if (ae.InnerException is Ice.LocalException)
    {
        _loggerAdmin.deadRemoteLogger(p, _localLogger,
                                      (Ice.LocalException)ae.InnerException, "log");
    }
}
```

Reply:
We should definitely remove Throwable from all UnknownException constructors and fix any code that use a UnknownException Throwable constructor.

The name Unknown is confusing but we can't change it. An UnknownException is not unknown. It's a very specific exception that means a dispatch failed with an "other" exception (other than local exception or user exception).
Whoever wrote the Java code you're showing above was unfortunately confused by the name.

We also don't want users to think "I'll throw an UnknownException that wraps some random exception - Ice likes that and will send something sensible to the client" because that's not the case.